File: smb_check.py
#!/usr/bin/env python3
# smb_check.py v3 — Comprueba y reconecta un recurso Samba 4 desde macOS

import subprocess
import logging
import sys
import time
from pathlib import Path

# ─── CONFIGURACIÓN ──────────────────────────────────────────
SMB_SERVER = "192.168.2.88"
SMB_SHARE = "2TB_BACKUP"
SMB_USER = "hans"
MOUNT_POINT = "/Volumes/2TB_BACKUP"
PROBE_FILE = MOUNT_POINT + "/.smb_probe"
LOG_FILE = Path.home() / "smb_monitor.log"
APP_NAME = "Monitor SMB"
# ────────────────────────────────────────────────────────────

logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
    handlers=[
        logging.FileHandler(LOG_FILE),
        logging.StreamHandler(sys.stdout),
    ],
)
log = logging.info


def run(cmd: list, timeout: int = 10) -> int:
    try:
        result = subprocess.run(
            cmd,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            timeout=timeout,
        )
        return result.returncode
    except (subprocess.TimeoutExpired, Exception):
        return -1


def run_output(cmd: list, timeout: int = 10) -> str:
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        return result.stdout.strip()
    except (subprocess.TimeoutExpired, Exception):
        return ""


def notify(title: str, message: str) -> None:
    applescript = (
        f'display notification "{message}" '
        f'with title "{APP_NAME}" '
        f'subtitle "{title}"'
    )
    try:
        subprocess.run(
            ["osascript", "-e", applescript],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            timeout=5,
        )
    except Exception:
        pass


def get_password() -> str:
    return run_output([
        "/usr/bin/security",
        "find-internet-password",
        "-a", SMB_USER,
        "-s", SMB_SERVER,
        "-w"
    ], timeout=5)


# ─── COMPROBACIONES ─────────────────────────────────────────

def check_server_reachable() -> bool:
    log("Comprobando conectividad con %s...", SMB_SERVER)
    if run(["/sbin/ping", "-c", "3", "-W", "2000", SMB_SERVER]) == 0:
        log("  OK - Servidor alcanzable")
        return True
    log("  ERROR - Servidor no responde al ping")
    return False


def check_smb_port() -> bool:
    log("Comprobando puerto SMB (445) en %s...", SMB_SERVER)
    if run(["/usr/bin/nc", "-z", "-w", "3", SMB_SERVER, "445"], timeout=8) == 0:
        log("  OK - Puerto 445 abierto")
        return True
    log("  ERROR - Puerto 445 cerrado o filtrado")
    return False


def is_mounted() -> bool:
    return MOUNT_POINT in run_output(["/sbin/mount"])


def is_writable() -> bool:
    log("  Comprobando escritura en el volumen...")
    rc = run(
        ["/bin/bash", "-c", f"echo ok > '{PROBE_FILE}' && cat '{PROBE_FILE}'"],
        timeout=5,
    )
    subprocess.run(["/bin/rm", "-f", PROBE_FILE],
                   stdout=subprocess.DEVNULL,
                   stderr=subprocess.DEVNULL)
    if rc == 0:
        log("  OK - Volumen accesible y con escritura")
        return True
    log("  ERROR - Volumen no responde a escritura")
    return False


# ─── RECONEXION ─────────────────────────────────────────────

def reconnect_smb() -> None:
    log("Intentando reconectar el recurso SMB...")

    mount_path = Path(MOUNT_POINT)
    if is_mounted() or mount_path.exists():
        log("  Desmontando recurso previo...")
        run(["/usr/sbin/diskutil", "unmount", "force", MOUNT_POINT])
        time.sleep(2)

    if mount_path.exists():
        log("  Punto de montaje sigue ocupado, forzando...")
        run(["/sbin/umount", "-f", MOUNT_POINT])
        time.sleep(1)

    # Obtener credenciales del Keychain
    password = get_password()
    if not password:
        log("  ERROR - No se encontro la contraseña en el Keychain")
        log(
            f'  Ejecuta: security add-internet-password -a "{SMB_USER}" -s "{SMB_SERVER}" -r "smb " -w "tu_contraseña"')
        notify("Error al reconectar",
               "No se encontro la contraseña en el Keychain")
        return

    # Abrir conexion via Finder con credenciales
    smb_url = f"smb://{SMB_USER}:{password}@{SMB_SERVER}/{SMB_SHARE}"
    log("  Abriendo conexion smb://%s:***@%s/%s...",
        SMB_USER, SMB_SERVER, SMB_SHARE)
    subprocess.Popen(["/usr/bin/open", smb_url])

    # Esperar hasta 20 segundos a que el montaje aparezca
    for _ in range(10):
        time.sleep(2)
        if is_mounted():
            log("  OK - Recurso reconectado correctamente")
            notify("Recurso reconectado",
                   f"{SMB_SHARE} vuelve a estar disponible")
            return

    log("  ERROR - No se pudo reconectar tras 20 segundos")
    notify("Error al reconectar",
           f"No se pudo montar {SMB_SHARE}. Revisa el log.")


# ─── EJECUCION PRINCIPAL ────────────────────────────────────

def main() -> None:
    log("====== Inicio de comprobacion SMB ======")

    if not check_server_reachable():
        notify("Sin conexion",
               f"No se puede alcanzar el servidor {SMB_SERVER}")
        log("Sin red, abortando")
        sys.exit(1)

    if not check_smb_port():
        notify("Puerto SMB cerrado",
               f"El puerto 445 de {SMB_SERVER} no responde")
        log("Puerto SMB cerrado, abortando")
        sys.exit(1)

    log("Comprobando estado del volumen %s...", MOUNT_POINT)

    if not is_mounted():
        log("  Volumen no montado, reconectando...")
        notify("Recurso desconectado",
               f"{SMB_SHARE} no esta montado - reconectando...")
        reconnect_smb()
    elif not is_writable():
        log("  Volumen montado pero no responde, reconectando...")
        notify("Recurso colgado", f"{SMB_SHARE} no responde - reconectando...")
        reconnect_smb()
    else:
        log("  Volumen montado y funcional, nada que hacer")

    log("====== Fin de comprobacion ======")


if __name__ == "__main__":
    main()

File: smb_check.js
#!/usr/bin/osascript -l JavaScript
// smb_check.js v10 — Comprueba y reconecta un recurso Samba 4 desde macOS (JXA)

// ─── CONFIGURACIÓN ──────────────────────────────────────────
const SMB_SERVER = "192.168.2.88";
const SMB_SHARE = "2TB_BACKUP";
const SMB_USER = "hans";
const MOUNT_POINT = "/Volumes/2TB_BACKUP";
const PROBE_FILE = MOUNT_POINT + "/.smb_probe";
const LOG_FILE = $.NSHomeDirectory().js + "/smb_monitor.log";
const APP_NAME = "Monitor SMB";
// ────────────────────────────────────────────────────────────

ObjC.import("Foundation");

// ─── HELPERS ────────────────────────────────────────────────

function timestamp() {
    const now = new Date();
    return now.toISOString().replace("T", " ").substring(0, 19);
}

function log(msg) {
    const line = `[${timestamp()}] ${msg}\n`;
    console.log(msg);
    const fm = $.NSFileManager.defaultManager;
    if (!fm.fileExistsAtPath($(LOG_FILE))) {
        $(line).dataUsingEncoding($.NSUTF8StringEncoding)
            .writeToFileAtomically($(LOG_FILE), true);
    } else {
        const fh = $.NSFileHandle.fileHandleForWritingAtPath($(LOG_FILE));
        fh.seekToEndOfFile;
        fh.writeData($(line).dataUsingEncoding($.NSUTF8StringEncoding));
        fh.closeFile;
    }
}

function run_cmd(args, timeout = 10) {
    const task = $.NSTask.alloc.init;
    const pipe = $.NSPipe.pipe;
    task.launchPath = $(args[0]);
    task.arguments = $(args.slice(1).map($));
    task.standardOutput = pipe;
    task.standardError = $.NSPipe.pipe;
    task.launch;

    const deadline = Date.now() + timeout * 1000;
    while (task.isRunning && Date.now() < deadline) {
        $.NSRunLoop.currentRunLoop
            .runUntilDate($.NSDate.dateWithTimeIntervalSinceNow(0.1));
    }
    if (task.isRunning) { task.terminate; return -1; }
    return task.terminationStatus;
}

function run_cmd_output(args, timeout = 10) {
    const task = $.NSTask.alloc.init;
    const pipe = $.NSPipe.pipe;
    task.launchPath = $(args[0]);
    task.arguments = $(args.slice(1).map($));
    task.standardOutput = pipe;
    task.standardError = $.NSPipe.pipe;
    task.launch;

    const deadline = Date.now() + timeout * 1000;
    while (task.isRunning && Date.now() < deadline) {
        $.NSRunLoop.currentRunLoop
            .runUntilDate($.NSDate.dateWithTimeIntervalSinceNow(0.1));
    }
    if (task.isRunning) { task.terminate; return ""; }

    const data = pipe.fileHandleForReading.readDataToEndOfFile;
    return $.NSString.alloc.initWithDataEncoding(data, $.NSUTF8StringEncoding).js.trim();
}

function sleep(seconds) {
    $.NSThread.sleepForTimeInterval(seconds);
}

function notify(title, message) {
    const app = Application.currentApplication();
    app.includeStandardAdditions = true;
    try {
        app.displayNotification(message, {
            withTitle: APP_NAME,
            subtitle: title,
        });
    } catch (_) { }
}

function getPassword() {
    return run_cmd_output([
        "/usr/bin/security",
        "find-internet-password",
        "-a", SMB_USER,
        "-s", SMB_SERVER,
        "-w"
    ], 5);
}

// ─── COMPROBACIONES ─────────────────────────────────────────

function checkServerReachable() {
    log("Comprobando conectividad con " + SMB_SERVER + "...");
    const rc = run_cmd(["/sbin/ping", "-c", "3", "-W", "2000", SMB_SERVER], 10);
    if (rc === 0) { log("  OK - Servidor alcanzable"); return true; }
    log("  ERROR - Servidor no responde al ping");
    return false;
}

function checkSmbPort() {
    log("Comprobando puerto SMB (445) en " + SMB_SERVER + "...");
    const rc = run_cmd(["/usr/bin/nc", "-z", "-w", "3", SMB_SERVER, "445"], 8);
    if (rc === 0) { log("  OK - Puerto 445 abierto"); return true; }
    log("  ERROR - Puerto 445 cerrado o filtrado");
    return false;
}

function isMounted() {
    const output = run_cmd_output(["/sbin/mount"], 5);
    return output.includes(MOUNT_POINT);
}

function isWritable() {
    log("  Comprobando escritura en el volumen...");
    const rc = run_cmd(
        ["/bin/bash", "-c", `echo ok > '${PROBE_FILE}' && cat '${PROBE_FILE}'`], 5
    );
    run_cmd(["/bin/rm", "-f", PROBE_FILE], 3);
    if (rc === 0) { log("  OK - Volumen accesible y con escritura"); return true; }
    log("  ERROR - Volumen no responde a escritura");
    return false;
}

// ─── RECONEXION ─────────────────────────────────────────────

function reconnectSmb() {
    log("Intentando reconectar el recurso SMB...");

    // Desmontar si esta montado o el directorio existe colgado
    const fm = $.NSFileManager.defaultManager;
    if (isMounted() || fm.fileExistsAtPath($(MOUNT_POINT))) {
        log("  Desmontando recurso previo...");
        run_cmd(["/usr/sbin/diskutil", "unmount", "force", MOUNT_POINT], 10);
        sleep(2);
    }

    // Segundo intento si el directorio sigue existiendo
    if (fm.fileExistsAtPath($(MOUNT_POINT))) {
        log("  Punto de montaje sigue ocupado, forzando...");
        run_cmd(["/sbin/umount", "-f", MOUNT_POINT], 5);
        sleep(1);
    }

    // Obtener credenciales del Keychain
    const password = getPassword();
    if (!password) {
        log("  ERROR - No se encontro la contraseña en el Keychain");
        log(`  Ejecuta: security add-internet-password -a "${SMB_USER}" -s "${SMB_SERVER}" -r "smb " -w "tu_contraseña"`);
        notify("Error al reconectar", "No se encontro la contraseña en el Keychain");
        return;
    }

    // Abrir conexion via Finder con credenciales
    const smb_url = `smb://${SMB_USER}:${password}@${SMB_SERVER}/${SMB_SHARE}`;
    log(`  Abriendo conexion smb://${SMB_USER}:***@${SMB_SERVER}/${SMB_SHARE}...`);
    run_cmd(["/usr/bin/open", smb_url], 5);

    // Esperar hasta 20 segundos a que el montaje aparezca
    let intentos = 0;
    while (intentos < 10) {
        sleep(2);
        if (isMounted()) {
            log("  OK - Recurso reconectado correctamente");
            notify("Recurso reconectado", `${SMB_SHARE} vuelve a estar disponible`);
            return;
        }
        intentos++;
    }

    log("  ERROR - No se pudo reconectar tras 20 segundos");
    notify("Error al reconectar", `No se pudo montar ${SMB_SHARE}. Revisa el log.`);
}

// ─── EJECUCION PRINCIPAL ─────────────────────────────────────

function main() {
    log("====== Inicio de comprobacion SMB ======");

    if (!checkServerReachable()) {
        notify("Sin conexion", `No se puede alcanzar el servidor ${SMB_SERVER}`);
        log("Sin red, abortando");
        return;
    }
    if (!checkSmbPort()) {
        notify("Puerto SMB cerrado", `El puerto 445 de ${SMB_SERVER} no responde`);
        log("Puerto SMB cerrado, abortando");
        return;
    }

    log("Comprobando estado del volumen " + MOUNT_POINT + "...");

    if (!isMounted()) {
        log("  Volumen no montado, reconectando...");
        notify("Recurso desconectado", `${SMB_SHARE} no esta montado - reconectando...`);
        reconnectSmb();
    } else if (!isWritable()) {
        log("  Volumen montado pero no responde, reconectando...");
        notify("Recurso colgado", `${SMB_SHARE} no responde - reconectando...`);
        reconnectSmb();
    } else {
        log("  Volumen montado y funcional, nada que hacer");
    }

    log("====== Fin de comprobacion ======");
}

main();